### PR TITLE
PHP: expose grpc_channel_credentials_copy() API from c-core

### PR DESCRIPTION
**Pull request by HannahShiSFB**
add 2 APIs in c-core to support php needs(put channel args and creds in persistent list, to prevent crash when folk)
related to #21321 and #21832

### CODE REVIEW

**Comment by stanley-cheung** [comment]
Sorry I think you misunderstood. According to the review comment at https://github.com/grpc/grpc/pull/21832#discussion_r395230619, let's re-use the existing `credentials_pointer_arg_copy` function, and expose it in grpc_security.h, instead of creating a new function.

**Comment by HannahShiSFB** [comment]
There are 4 reasons I didn't use credentials_pointer_arg_copy and make it to an API as the reviewer suggested.
1. It's an intra function, not designed to be used as an API. 
2. It doesn't have a type, void* is not good enough for an API.
3. It is paired with another intra function credentials_pointer_arg_destroy.
4. Our API supposes to pair with another existing API grpc_channel_credentials_release.

**Comment by stanley-cheung** [comment]
@ZhenLian We have separated the piece that we need from #21832 and put it here in this separate PR. We still have some reasons (see comment above) that we felt that a new function is more appropriate. Can you please weigh in? Thanks!

File: src/core/lib/security/credentials/credentials.cc [hunk 1]
```diff
@@ -45,6 +45,14 @@ void grpc_channel_credentials_release(grpc_channel_credentials* creds) {
   if (creds) creds->Unref();
 }
 
+grpc_channel_credentials* grpc_channel_credentials_copy(
+    grpc_channel_credentials* creds) {
+  GRPC_API_TRACE("grpc_channel_credentials_copy(creds=%p)", 1, (creds));
+  grpc_core::ExecCtx exec_ctx;
```
Comment (jiangtaoli2016) [inline src/core/lib/security/credentials/credentials.cc:51]: Why do we need a new exec_ctx?

File: src/core/lib/security/credentials/credentials.cc [hunk 2]
```diff
@@ -45,6 +45,14 @@ void grpc_channel_credentials_release(grpc_channel_credentials* creds) {
...
+  if (creds) creds->Ref().release();
```
Comment (jiangtaoli2016) [inline src/core/lib/security/credentials/credentials.cc:52]: It is odd to return the original creds.
You can either make the function void or return the copied object, e.g., 
` if (creds) return creds->Ref().release();`

File: src/core/lib/security/credentials/credentials.cc (hunk 1 above)
Comment (HannahShiSFB) [reply src/core/lib/security/credentials/credentials.cc:51]: This is to be consistent with other similar functions, like grpc_channel_credentials_release() and the ctx will have the call-stack, good for debugging purposes.

File: src/core/lib/security/credentials/credentials.cc (hunk 2 above)
Comment (HannahShiSFB) [reply src/core/lib/security/credentials/credentials.cc:52]: This is to be consistent with other similar functions, like channelz_node_copy() and target_authority_table_copy().

**Comment by jiangtaoli2016** [comment]
@markdroth For c core API change, do we need a gRFC?

**Comment by HannahShiSFB** [comment]
if I may comment on the whole picture of why grpc-php-ext needs c-core's support

In #21321, a google cloud client raised an issue: “Subprocess crash (via system) after simple rpc call attempt”. We reproduced and analysed. It turned out that grpc-php-ext needs to hold a copy of channel args and creds itself.

Here is why:
Channel args and creds are created and their pointers(references) are stored in the persistent list, when a channel(any channel, no matter it’s parent process or subprocess/fork) is created by request. And they will all be destroyed by php when the request ends, since the creds and part of the channel args struct are php objects. As a result, from this point, the pointers(references) in the persistent list became dangling pointers.
When grpc-php-ext uses these pointers(references) after, the process crashes.

Propose to fix:
grpc-php-ext makes a copy of the entire channel args struct and creds and manages itself.

What support grpc-php-ext needs from c-core:
1. An API to copy creds, since grpc-php-ext doesn’t know how to do it itself.
2. Expose grpc_channel_args_copy() and grpc_channel_args_destroy() to allow grpc-php-ext to copy channel args, otherwise grpc-php-ext needs to write the same code c-core already has.

**Comment by markdroth** [comment]
> Channel args and creds are created and their pointers(references) are stored in the persistent list, when a channel(any channel, no matter it’s parent process or subprocess/fork) is created by request. And they will all be destroyed by php when the request ends, since the creds and part of the channel args struct are php objects. As a result, from this point, the pointers(references) in the persistent list became dangling pointers.
> When grpc-php-ext uses these pointers(references) after, the process crashes.

I don't think I know enough about PHP to understand this explanation.  I suggest that someone schedule a meeting to discuss this in more detail.

**Comment by stanley-cheung** [comment]
> > Channel args and creds are created and their pointers(references) are stored in the persistent list, when a channel(any channel, no matter it’s parent process or subprocess/fork) is created by request. And they will all be destroyed by php when the request ends, since the creds and part of the channel args struct are php objects. As a result, from this point, the pointers(references) in the persistent list became dangling pointers.
> > When grpc-php-ext uses these pointers(references) after, the process crashes.
> 
> I don't think I know enough about PHP to understand this explanation. I suggest that someone schedule a meeting to discuss this in more detail.

@markdroth Thanks for the feedback. I understand that there might be reason that c-core might not want to expose the channel args struct. We have updated this PR to only ask for `grpc_channel_credentials_copy()`. We can try to save the channel args only in PHP code. Would that be OK? Thanks.

**Comment by markdroth** [comment]
I would still like to have a meeting to discuss this.  I am not comfortable approving any C-core API changes without better understanding why they are necessary.

**Comment by stanley-cheung** [comment]
> I would still like to have a meeting to discuss this. I am not comfortable approving any C-core API changes without better understanding why they are necessary.

Sounds good. I will put together the details and set up a meeting. Thanks!

**Comment by stanley-cheung** [comment]
Superceded by #22774

**Pull request closed without merging**